### ISSUE TITLE
Infinite loops in CD_Geometry

### ISSUE THREAD

**Issue by randlet** [nrc-cnrc/EGSnrc#97]
We have a problem where very rarely (i.e.  ~ 1/10E9 histories) the code is getting stuck in an infinite loop in [this code block in CD_Geometry](https://github.com/nrc-cnrc/EGSnrc/blob/master/HEN_HOUSE/egs%2B%2B/geometry/egs_cd_geometry/egs_cd_geometry.h#L611-L687).    This occurs less frequently after the fixes from @mchamberland's CD geom pull requests are applied but still is causing us some issues.

We don't have a test case where it's fast to replicate the issue yet but I'm assuming this is a floating point /  roundoff issue.  

Ideally we could prevent this from ever happening but given the rarity of the issue I'm wondering if we can just terminate the loop (output a warning?) after a certain number of iterations and return e.g. whatever region in the base geometry we are "in" (ibase_new).  Not sure what the best approach is, but IMO it's bad practice to have loops which may not ever terminate so at the very least I believe there should be an iteration counter (or switch the loop to a for loop) which terminates the simulation if it gets stuck.

**Comment by rtownson** [comment]
Does anyone have an input file they can share where this occurs?

@blakewalters, do you think we can replace the `while(1)` with a `for` loop like the one below? This lets the simulation continue, but we can give warnings that their simulation will be slow as a result. Should the `1e6` be smaller or larger?

````
for (EGS_I64 i=0; i<1e6; i++) {
  ...
  if(i == 1e6-1) egsWarning("Infinite loop avoided...");
}
````

**Comment by mchamberland** [comment]
@rtownson We came upon this issue in our custom application, unfortunately.

@randlet Can you reproduce this in `tutor7pp` if you simply copy the geometry and source sections of the input file (assuming you still have it)?

**Comment by blakewalters** [comment]
I don't have a problem with this...as long as it doesn't replace figuring out what's actually going on!

**Comment by randlet** [comment]
The one good test case that I had (would freeze egs_view almost instantly) actually seems to have been resolved by @mchamberland's  23763c2ea04e52 so it's possible it was more of an issue with the cone geometry rather than the CD.

**Comment by randlet** [comment]
That said, I still think it's worthwhile having a sanity check in there like the one @rtownson suggested.
